lib/Infer/InstSynthesis.cpp

```diff
@@ -0,0 +1,1264 @@
+      if (ProgramWiring.count(Line)) {
+        // Look for the output wiring. It can be a component's output or any
+        // of the inputs. Note that due to scalability reasons, we haven't
+        // encoded width mismatches and other invalid wirings as constraints,
+        // thus, we need to perform sanity checks on candidates
+        if (!OutLocSet) {
+          // Is it output? If yes, there should be only one component output
+          // and/or any input on that line
+          auto const &VarMap = ProgramWiring[Line];
+          if (Loc == O.first) {
+            for (auto const &CandLocVar : VarMap) {
+              if (!isWiringInvalid(O.first, CandLocVar)) {
+                OutLocVar = CandLocVar;
+                OutLocSet = true;
+                break;
+              }
+            }
+          // Check if the output is stored already. If yes, then take the new
+          // location if the width matches
+          } else if (VarMap.count(O.first) && !isWiringInvalid(O.first, Loc)) {
+            OutLocVar = Loc;
+            OutLocSet = true;
+          }
+        }
+        ProgramWiring[Line].insert(Loc);
+      } else {
+        ProgramWiring[Line] = {Loc};
+      }
+      Counter++;
+    // Parse constant models
+    } else if (Name.find(CONST_PREFIX) != std::string::npos) {
+      LocVar Loc = getLocVarFromStr(Name.substr(CONST_PREFIX.size()));
+      ConstValMap[Loc] = ModelVals[J];
+    }
+  }
+  assert(OutLocSet && "no matching location for the output");
+  assert(ProgramWiring.size() <= M && "the output location must be <= M");
+  assert(Counter == L.size() && "invalid number of locations in the model");
+
+  return OutLocVar;
+}
+
+LocVar InstSynthesis::getWiringLocVar(const LocVar &OpLoc,
+                                      const LineLocVarMap &ProgramWiring) {
+  LocVar Match;
+  bool FoundMatch = false;
+
+  if (DebugSynthesis)
+    llvm::outs() << "- looking for OpLoc wiring "
+                 << getLocVarStr(OpLoc) << "\n";
+  for (auto const &E : ProgramWiring) {
+    if (E.second.count(OpLoc)) {
+      if (DebugSynthesis)
+        llvm::outs() << "- found wiring input on line " << E.first << ", taking ";
+      for (auto const &In : E.second) {
+        // Take either input, constant, or component output of matching width
+        if ((In.first == 0 || In.second == 0) && !isWiringInvalid(In, OpLoc)) {
+          Match = In;
+          if (DebugSynthesis)
+            llvm::outs() << getLocVarStr(Match);
+          FoundMatch = true;
+          break;
+        }
+      }
+      if (DebugSynthesis)
+        llvm::outs() << "\n";
+    }
+    if (FoundMatch)
+      break;
+  }
+
+  return Match;
+}
+
+Inst *InstSynthesis::createJunkFreeInst(Inst::Kind Kind, unsigned Width,
+                                        std::vector<Inst *> &Ops,
+                                        InstContext &IC) {
+  switch (Kind) {
+  case Inst::Add:
+  case Inst::AddNSW:
```

Comment:
But aren't the nsw/nuw variants in the list of components not supported by synthesis above?